types/document.d.ts

```diff
@@ -90,9 +90,9 @@ declare module 'mongoose' {
     $session(session?: ClientSession | null): ClientSession | null;
 
     /** Alias for `set()`, used internally to avoid conflicts */
-    $set(path: string, val: any, type: any, options?: DocumentSetOptions): this;
-    $set(path: string, val: any, options?: DocumentSetOptions): this;
-    $set(value: any): this;
+    $set(path: string | Record<string, any>, val: any, type: any, options?: DocumentSetOptions): this;
```

Comment:
This isn't quite correct, Mongoose ignores `val` if `path` is an object. So I think the more correct approach would be to add an additional override to `$set()` that looks like this:

```
$set(Record<string, any>, val?: null | undefined, type?: any, options?: DocumentSetOptions): this;
```

What do you think?

Reply:
> This isn't quite correct, Mongoose ignores val if path is an object. 

if i understand it correctly, a additional overload would be needed like

```ts
$set(path: Record<string, any>, type: any, options?: DocumentSetOptions): this;
```

or is `type` not just assigned from parameter 2 and still needs `val: never`?